airbyte-cdk/python/unit_tests/base_python/test_http.py

```diff
@@ -0,0 +1,130 @@
+
+    url_base = "https://test_base_url.com"
+    primary_key = ""
+
+    def __init__(self):
+        super().__init__()
+        self.resp_counter = 1
+
+    def next_page_token(self, response: requests.Response) -> Optional[Mapping[str, Any]]:
+        return None
+
+    def path(
+        self, stream_state: Mapping[str, Any] = None, stream_slice: Mapping[str, Any] = None, next_page_token: Mapping[str, Any] = None
+    ) -> str:
+        return ""
+
+    def parse_response(
+        self,
+        response: requests.Response,
+        stream_state: Mapping[str, Any],
+        stream_slice: Mapping[str, Any] = None,
+        next_page_token: Mapping[str, Any] = None,
+    ) -> Iterable[Mapping]:
+        stubResp = {"data": self.resp_counter}
+        self.resp_counter += 1
+        yield stubResp
+
+
+def test_stub_basic_read_http_stream_read_records(mocker):
+
+    stream = StubBasicReadHttpStream()
+    blank_response = {}  # Send a blank response is fine as we ignore the response in `parse_response anyway.
+    mocker.patch.object(StubBasicReadHttpStream, "_send_request", return_value=blank_response)
+
+    records = []
+    for r in stream.read_records(SyncMode.full_refresh):
+        records.append(r)
+
+    assert [{"data": 1}] == records
+
+
+class StubNextPageTokenHttpStream(StubBasicReadHttpStream):
+
+    current_token = 0
+    max_token_number = 6
+
+    def next_page_token(self, response: requests.Response) -> Optional[Mapping[str, Any]]:
+        while self.current_token < 5:
+            self.current_token += 1
+            return {"token": "token"}
+        return None
+
+
+def test_stub_next_page_token_http_stream_read_records(mocker):
+
+    stream = StubNextPageTokenHttpStream()
+    blank_response = {}  # Send a blank response is fine as we ignore the response in `parse_response anyway.
+    mocker.patch.object(StubNextPageTokenHttpStream, "_send_request", return_value=blank_response)
+
+    records = []
+    for r in stream.read_records(SyncMode.full_refresh):
+        records.append(r)
+
+    assert [{"data": 1}, {"data": 2}, {"data": 3}, {"data": 4}, {"data": 5}, {"data": 6}] == records
+
+
+class StubBadUrlHttpStream(StubBasicReadHttpStream):
+
+    url_base = "bad_url"
+
+
+def test_stub_bad_url_http_stream_read_records(mocker):
+
+    stream = StubBadUrlHttpStream()
+
+    with pytest.raises(requests.exceptions.RequestException):
+        records = []
+        for r in stream.read_records(SyncMode.full_refresh):
+            records.append(r)
+
+
+class StubShouldNotRetryHttpStream(StubBasicReadHttpStream):
+    def backoff_time(self, response: requests.Response) -> Optional[float]:
+        return 0.5
+
+
+def test_1(mocker):
+
+    stream = StubShouldNotRetryHttpStream()
+    req = requests.Response()
+    req.status_code = 429
+
+    mocker.patch.object(requests.Session, "send", return_value=req)
+
+    with pytest.raises(UserDefinedBackoffException):
+        records = []
```

Comment:
I was trying to spy on `_send_request` and assert the number of calls, but this it's always '1'. I'm going to guess this is because of the way the annotations are set up?

In any case, I can see from logs retries are happening so I'm going to leave a TODO for now, and move on. What do you think?

Reply:
https://github.com/airbytehq/airbyte/issues/3306